### PR TITLE
[PROD-678] Authorization and password reset vulnerability fix

### PR DESCRIPTION
**Pull request by AndrewVolosytnykhThingsboard**

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/controller/AuthController.java [hunk 1]
```diff
@@ -163,8 +163,12 @@ public void requestResetPasswordByEmail (
             @RequestBody JsonNode resetPasswordByEmailRequest,
             HttpServletRequest request) throws ThingsboardException {
         try {
+            mailService.testConnection();
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/AuthController.java:166]: Why test connection on each reset password?

File: dao/src/main/java/org/thingsboard/server/dao/user/UserServiceImpl.java [hunk 2]
```diff
@@ -194,11 +194,13 @@ public UserCredentials requestPasswordReset(TenantId tenantId, String email) {
         DataValidator.validateEmail(email);
         User user = userDao.findByEmail(tenantId, email);
         if (user == null) {
-            throw new IncorrectParameterException(String.format("Unable to find user by email [%s]", email));
+            log.error(String.format("Unable to find user by email [%s]", email));
```
Comment (ashvayka) [inline dao/src/main/java/org/thingsboard/server/dao/user/UserServiceImpl.java:197]: Why not throw an appropriate exception? Return null in such a case is bad practice.

File: application/src/main/java/org/thingsboard/server/controller/AuthController.java (hunk 1 above)
Comment (AndrewVolosytnykhThingsboard) [reply application/src/main/java/org/thingsboard/server/controller/AuthController.java:166]: @ashvayka I making test connection to hide exceptions when we specify correct email (present in database) and block execution of code if mail server not configured or not available, to throw more user-friendly message like 'Unable to connect to mail server' not 'Unable to send mail: 535-5.7.8 Username and Password not accepted. Learn more at 535 5.7.8 https://support.google.com/mail/?p=BadCredentials g2sm1957632ljn.35 - gsmtp'

File: application/src/main/java/org/thingsboard/server/controller/AuthController.java [hunk 3]
```diff
@@ -73,6 +75,7 @@
     private final SystemSecurityService systemSecurityService;
     private final AuditLogService auditLogService;
     private final ApplicationEventPublisher eventPublisher;
+    private final Executor executor = Executors.newSingleThreadExecutor();
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/AuthController.java:78]: is this ever closed?

Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/controller/AuthController.java:78]: Use MailExecutorService

File: application/src/main/java/org/thingsboard/server/controller/AuthController.java [hunk 4]
```diff
@@ -159,24 +162,26 @@ public UserPasswordPolicy getUserPasswordPolicy() throws ThingsboardException {
 
     @RequestMapping(value = "/noauth/resetPasswordByEmail", method = RequestMethod.POST)
     @ResponseStatus(value = HttpStatus.OK)
-    public void requestResetPasswordByEmail (
+    public void requestResetPasswordByEmail(
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/AuthController.java:165]: Add sendResetPasswordEmailAsync to the Mail Service

File: application/src/main/java/org/thingsboard/server/controller/AuthController.java [hunk 5]
```diff
@@ -75,11 +73,11 @@
     private final SystemSecurityService systemSecurityService;
     private final AuditLogService auditLogService;
     private final ApplicationEventPublisher eventPublisher;
-    private final Executor executor = Executors.newSingleThreadExecutor();
 
     @PreAuthorize("isAuthenticated()")
     @RequestMapping(value = "/auth/user", method = RequestMethod.GET)
-    public @ResponseBody User getUser() throws ThingsboardException {
+    public @ResponseBody
+    User getUser() throws ThingsboardException {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/AuthController.java:80]: why new line?

**Pull request merged**